src/man/man3/sam_clear_request.3
.\" $Revision: 1.17 $
.ds ]W Sun Microsystems
.\" SAM-QFS_notice_begin
.\"
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
.\" or https://illumos.org/license/CDDL.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at pkg/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.\"
.\" Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
.\" Use is subject to license terms.
.\"
.\" SAM-QFS_notice_end
.TH sam_clear_request 3 "02 Jun 2004"
.SH NAME
sam_clear_request \- Clears entry from removable media mount requests
.SH SYNOPSIS
.LP
.BI "cc [ " "flag"
.BI " ... ] " "file"
.BI " ... -L/opt/SUNWsamfs/lib -lsamapi [ " "library" " ... ]"
.LP
.nf
.ft 3
#include "/opt/SUNWsamfs/include/samapi.h"
.ft
.fi
.LP
.BI "int sam_clear_request(uint_t " "slot" ,
.BI "int " "wait_response" );
.SH DESCRIPTION
.B sam_clear_request(\|)
clears removable media entry from mount request list.  The call will return
immediately after issuing the command if zero is specified for
.I wait_response
value.  Other values for
.I wait_response
will give undefined results.
.SH "RETURN VALUES"
Upon succesful completion a value of 0 is returned.
Otherwise, a value of \-1 is returned and
\f4errno\fP
is set to indicate the error.
.SH ERRORS
.PP
.B sam_clear_request(\|)
fails if one or more of the following are true:
.TP 20
.SB ER_FIFO_PATH_LENGTH
The path and filename for the FIFO pipe is too long.
.TP
.SB ER_NO_MASTER_SHM
No Sun StorEdge \%SAM-FS or Sun \%SAM-QFS master shared memory segment defined.
Check that the Sun StorEdge \%SAM-FS or Sun \%SAM-QFS file systems are mounted.
.TP
.SB ER_NO_MASTER_SHM_ATT
No Sun StorEdge \%SAM-FS or \%SAM-QFS master shared memory segment found.
Check that the Sun StorEdge \%SAM-FS or Sun \%SAM-QFS file systems are mounted.
.TP
.SB ER_NO_RESPONSE_FIFO
Unable to create the response FIFO pipe.
.TP
.SB ER_OPERATOR_NOT_PRIV
Operator does not have permission to perform a full audit
.SH SEE ALSO
.BR samu (8).


src/lib/samapi/sam_clear_request.c
/*
 * sam_clear_request.c - Clear the specified slot from the mount requests
 *
 *	sam_clear_request() function clears the specified slot or entry from
 *	the removable media mount requests
 *
 */


/*
 *    SAM-QFS_notice_begin
 *
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
 * or https://illumos.org/license/CDDL.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at pkg/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 *
 *    SAM-QFS_notice_end
 */

#pragma ident "$Revision: 1.20 $"


#include <stdio.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>
#include <sys/param.h>
#include <sys/types.h>
#include <sys/shm.h>
#include <unistd.h>

#include "sam/types.h"
#include "sam/param.h"
#include "aml/device.h"
#include "sam/defaults.h"
#include "aml/fifo.h"
#include "sam/lib.h"
#include "aml/proto.h"
#include "aml/sam_utils.h"
#include "aml/shm.h"

/* API headers. */
#include "aml/samapi.h"
#include "pub/sam_errno.h"




/*
 *	sam_clear_request() - API function to clear a removable media mount
 *			      request
 *
 *	Input parameters --
 *		slot		Slot number
 *		wait_response	Wait for command response if nonzero
 *				-1	Wait forever
 *				> zero	Wait this many seconds
 *					for response
 *
 *	Output parameters --
 *		errno		Set to error number if error encountered
 *
 *	Return value --
 *		0		If successful in moving the media
 *		-1		If error encountered,
 *				'errno' set to error number
 *
 */

int
sam_clear_request(
	uint_t	 slot,		/* Slot number */
	int wait_response)	/* Set nonzero to wait for response */
{
	char	*fifo_path;	/* Path to FIFO pipe, from shared memory */
	int	cmd = CMD_FIFO_DELETE_P;

	operator_t	operator;	/* Operator data */
	sam_cmd_fifo_t	cmd_block;
	shm_alloc_t	master_shm;	/* Master shared memory structure */


	/*
	 *	Access device table in shared memory segment.
	 */

	if ((master_shm.shared_memory =
	    (shm_ptr_tbl_t *)sam_mastershm_attach(0, SHM_RDONLY)) ==
	    (void *) -1) {
		return (-1);
	}


	/*
	 *	Check if operator has privilege for label
	 */

	SET_SAM_OPER_LEVEL(operator);

	if (!SAM_OPER_CLEAR(operator)) {
		shmdt((char *)master_shm.shared_memory);
		errno = ER_OPERATOR_NOT_PRIV;
		return (-1);
	}

	fifo_path = (char *)SHM_REF_ADDR(((shm_ptr_tbl_t *)
	    master_shm.shared_memory)->fifo_path);

	/*
	 * Clear command block
	 */

	memset(&cmd_block, 0, sizeof (cmd_block));

	/*
	 *	Format up commmand to sam-amld
	 */

	cmd_block.magic = CMD_FIFO_MAGIC;
	cmd_block.slot = slot;
	cmd_block.cmd = cmd;

	/*
	 *	Send commmand to sam-amld
	 */

	if (sam_send_cmd(&cmd_block, wait_response, fifo_path) < 0) {
		shmdt((char *)master_shm.shared_memory);
		return (-1);
	}

	shmdt((char *)master_shm.shared_memory);
	return (0);
}
